README.md
# Ruby/Rails OmniAuth for HumanID Alpha

Omniauth for humanID, a platform that prevents bots and increases privacy. HumanID is run by Human Internet,
a non-profit that is currently financed by organizations such as Harvard and the Mozilla Foundation (I love the Mozilla
Developer Network (MDN) which gives great javascript information).

HumanID works best when used as the only sign-up solution, due to this HumanID has to be highly trusted. This is where their
non-profit status steps in. HumanID has many benifits:

1. Increased privacy for users through both technical innovations and legal responsibilities.
2. Making bots inconvienient by requiring phone verification.
3. 'One voice one vote' type benefits by making it difficult to have multiple accounts.
4. Dont have to deal with users putting in "password123" as a password.
5. Dont have to deal with email/password registration generally (my signup process became way less complicated, there was all this stuff which to be honest I just didn't really get).
6. Much better look than "sign up with [tech monopoly here]" buttons.

## Installation

Add this line to your application's Gemfile:

```ruby
gem 'omniauth-humanid'
```

And then execute:

    $ bundle install

Or install it yourself as:

    $ gem install omniauth-humanid
    
Update as normal.

## Configuration

1. Make an account at humanID, and get appropriate credentials
2. You will need to set credentials for client-id and client-secret, for both development and production. This can be done using the rails:credentials method or Enviroment Variables. These both have various advantage/disadvantages.
3. do some configuration in your initializers, for devise the steps are:
    - open config/initializers/devise.rb
    - enter the below code within the "Devise.setup" area
    ```ruby
    Devise.setup do |config|
        id = Rails.application.credentials.omniauth_humanid_client_id
	    secret = Rails.application.credentials.omniauth_humanid_client_secret
	    config.omniauth :humanid, client_secret: secret, client_id: id
    end
    ```
4. Make your signup button
    - I use html which is a bit diffrent than what was provided to allow for CSRF concerns raised in [this gem](https://github.com/cookpad/omniauth-rails_csrf_protection). To follow these instructions, first install and setup that gem
    - Next your going to want to put this button code in a partial since you will be using it semi-regularly in your code base (convert to irb if needed):
    ```ruby
    .btn.common-sign-up-with-btn
	= form_with url: user_humanid_omniauth_authorize_path, method: :post do
		%input{type: :image, src: image_pack_path("icons/sign_in_logos/humanID.svg"), alt: "Anonymous Login with humanID"}
    ```
5. Create your callback area (still in development)
    - This area is generally supposed to be customizable, as you might have a diffrent model name, want to attach some validations, etc, etc. So it is not included in the gem, but is here as a how-to.
    - TBD

## Additional configuration

### Omniauth options

additional configuration can be set in your initializer file at the same area and in the same method as your client-secret and client-id. Additional configuration is as follows: 
- lang: default language code. Defaults to 'en' (english). Set to nil to remove from url.
- humanid_version: version string that goes in the url. Defaults to 'v0.0.3'. If humanid updates this may need to be updated aswell.
- priority_country: not sure exactly what this does or how to use it, but it was in the docs so i added it as an option. Defaults to nil.
- external_signup_url: the web login url. Defaults to: "https://core.human-id.org/[HUMANID_VERSION]/server/users/web-login". [HUMANID_VERSION] gets substituted by humanid_version above.

### Devise without emails/passwords

Once again, humanID is better when used alone bue to bot mitigation. This section is how to remove email/password authentication.

Although Devise is easier to deal with without usernames / passwords, it takes a bit to get there. This is out of the gem scope, but here are some pointers below. You may run into other hurdles depending on your setup.
1. In your user.rb (or similar) model, in your devise config line, remove all of the following:
	- confirmable
	- database_authenticatable
	- recoverable
	- confirmable
	- lockable
2. In your devise.rb initializer file, make sure to set authentication_keys to []
3. delete or comment out the selections in devise.rb related to number 1.
4. I had to add back the route below:
```ruby
as :user do
	delete "/users/sign_out" => "users/sessions#destroy"
end
```
5. For development you may have to create a seperate way to login/signup for testing purposes. You can do this by sending a form that implements the method 'sign_in_and_redirect user, event: :authentication', or that sets fake values for signup. MAKE SURE THIS METHOD IS ONLY ACTIVE DURING DEVELOPMENT. I have a version of this below:
	- in my routes.rb:
	```ruby
	as :user do
		if Rails.env.development?
			post '/users/override' => 'users/omniauth_callbacks#callback_override'
		end
	end
	```
	- in my OmniauthCallbacksController override (see devise documentation):
	```ruby
	def callback_common(provider, uid)
		user = User.from_omniauth(provider, uid)
		if user
			#allready have an account, sign them in
			sign_in_and_redirect user, event: :authentication # this will throw if user is not activated
		else
			request.session['signup'] ||= {}
			request.session["signup"]["provider"] = provider
			request.session["signup"]["uid"] = uid
			raise StandardError.new("REPLACE THIS ERROR WITH A REDIRECT TO FINISH SIGNUP")
		end
	end
	if Rails.env.development?
		def callback_override
			raise StandardError.new("nope") unless Rails.env.development?
			provider = 'override'
			uid = params['uid']
			callback_common provider, uid
		end
	end
	```
	- my form (hidden in a dropdown menu)
	```ruby
		- if Rails.env.development?
			.dropdown-item
				= form_with url: users_override_path, method: :post do
					%p put username below to bypass the humanID in development
					= text_field_tag :uid, '', class: 'form-control'
					= submit_tag "GO", class: "btn btn-primary"
	```

## Development of the Gem (Gem usage info stops here)

After checking out the repo, run `bin/setup` to install dependencies.

To install this gem onto your local machine, run `bundle exec rake install`. To release a new version, update the version number in the gemspec file, and then run `bundle exec rake release`, which will create a git tag for the version, push git commits and tags, and push the `.gem` file to [rubygems.org](https://rubygems.org).

When developing the Gem, for minor updates while debugging etc, we have a custom script to update the gem
- in the root folder of the gem run ./bin/update. This will bump the version number by one in the gemspec, commit all changes with a generic message, and then install this minor version on your machine.
- For more information see the ./bin/update file.

## Contributing

Bug reports and pull requests are welcome on GitHub at https://github.com/LukeClancy/omniauth-humanid. This project is intended to be a safe, welcoming space for collaboration, and contributors are expected to adhere to the [code of conduct](https://github.com/LukeClancy/omniauth-humanid/blob/master/CODE_OF_CONDUCT.md).

## License

The gem is available as open source under the terms of the [MIT License](https://opensource.org/licenses/MIT).

## Code of Conduct

Everyone interacting in the Omniauth::Humanid project's codebases, issue trackers, chat rooms and mailing lists is expected to follow the [code of conduct](https://github.com/LukeClancy/omniauth-humanid/blob/master/CODE_OF_CONDUCT.md).


changelog.md
## This page will have proper updates when it is functional enough to warrent that


lib/omniauth/strategies/humanid.rb
require 'omniauth'
require 'json'

module OmniAuth
	module Strategies
		class Humanid
			include OmniAuth::Strategy
			#Omniauth strategy creation guide be useful
			#https://github.com/omniauth/omniauth/wiki/Strategy-Contribution-Guide

			#then this guy created another useful blog 🔥🔥🔥🔥🔥🔥🔥🔥🔥🔥🔥🔥🔥🔥🔥🔥🔥🔥🔥🔥🔥🔥🔥🔥🔥🔥🔥🔥🔥🔥🔥🔥🔥🔥🔥🔥🔥🔥🔥🔥🔥🔥🔥🔥🔥🔥🔥🔥🔥🔥🔥🔥🔥🔥🔥🔥🔥🔥🔥🔥🔥🔥🔥🔥🔥🔥🔥🔥🔥🔥🔥🔥🔥🔥🔥🔥🔥🔥🔥🔥🔥🔥🔥🔥🔥🔥🔥🔥🔥🔥🔥🔥🔥🔥🔥🔥🔥🔥🔥🔥🔥🔥
			#https://dev.to/vvo/devise-create-a-local-omniauth-strategy-for-slack-4066

			#note the image in the below documentation, I will try to reference back to it below
			#https://docs.human-id.org/web-sdk-integration-guide#api-request-web-log-in-session

			provider :humanid
			options :login_button_path, '[LOGIN_BUTTON_PATH]'
			options :humanid_version, 'v0.0.3'
			options :local_sign_up_url, "/omniauth/humanid"
			options :external_signup_url, "https://core.human-id.org/[HUMANID_VERSION]/server/users/web-login"
			options :lang, :en
			options :client_secret, nil
			options :client_id, nil

			def self.humanid_button
				#see https://docs.human-id.org/web-sdk-integration-guide
				%Q{<a href="#{options.local_sign_up_url}">
    				<img src="#{options.login_button_path}" alt="Anonymous Login with humanID" height="27"/>
				<a>}
			end

			def get_client_id
				return options.client_secret unless options.client_secret.nil?
				raise StandardError.new("Set humanid client id")
			end
			def get_client_secret
				return options.client_secret unless options.client_secret.nil?
				raise StandardError.new("Set humanid client secret")
			end
			def get_external_signup_url
				url = options.external_signup_url.gsub('[HUMANID_VERSION]', options.humanid_version)
				url += "?lang=#{options.lang}" if options.lang
			end
			def request_phase_err(res)
				raise StandardError.new("Issue with the request phase of humanid omniauth, response from human id has code: #{res.code}, and body: #{res.body}")
			end
			#request phase
			def request_phase
				# In the humanid web-sdk-integration-guide, this would be the "[1] login" step. We need to get the redirect url
				# through a post request, and then send that to the user.

				#see more on Net::HTTP here:
				#	https://ruby-doc.org/stdlib-2.4.1/libdoc/net/http/rdoc/Net/HTTP.html

				#get uri
				uri = URI(get_external_signup_url)
				#make a post request (but dont send it yet)
				post_request = Net::HTTP::Post.new(uri)
				#set the headers as per docs.
				post_request['client-id'] = get_client_id
				post_request['client-secret'] = get_client_secret
				post_request['Content-Type'] = 'application/json'
				#send the request using a weirdly 🤷‍♂️ overcomplicated 🤷‍♂️ method 🤷‍♂️ and 🤷‍♂️ block 🤷‍♂️ blame Net::HTTP
				res = Net::HTTP.start(uri.hostname, uri.port, use_ssl: true){|http| http.request(post_request)}

				if res.code == "200"
					body = Json.parse(res.body)
					# they have five diffrent metrics for success:
					#	1. the response code 200
					#	2. body["success"] == true
					#	3. body["code"] == "OK"
					#	4. body["message"] == "success"
					#	5. body["data"]["webLoginUrl"] is actually there
					# I check 1 and 5 since the others seem supplimentary
					
					#get the redirect url
					redirect_url = body.dig("data", "webLoginUrl")
					#check it, throw an error if nil
					request_phase_err(res) if redirect_url.nil?
					#redirect (everything is working!)
					redirect redirect_url
				else
					request_phase_err(res)
				end
			end
		end
	end
end


lib/omniauth-humanid.rb
require 'json'

module OmniAuth
	module Strategies
		class Humanid
			include OmniAuth::Strategy
			#Omniauth strategy creation guide be useful
				#- https://github.com/omniauth/omniauth/wiki/Strategy-Contribution-Guide
				#- note the request_phase and the callback_phase

			#note the image in the below documentation, I will try to reference back to it below
			#https://docs.human-id.org/web-sdk-integration-guide#api-request-web-log-in-session

			#options :login_button_path, '[LOGIN_BUTTON_PATH]'
			option :humanid_version, 'v0.0.3'
			#options :local_sign_up_url, "/auth/humanid"
			option :external_signup_url, "https://core.human-id.org/[HUMANID_VERSION]/server/users/web-login"
			option :lang, :en
            option :priority_country, nil #this is an option in the docs, but they dont give an example value (otherwise I would set to united_states, or us, or 1)
			option :client_secret, nil
			option :client_id, nil
			option :exchange_url, "https://core.human-id.org/[HUMANID_VERSION]/server/users/exchange"

			attr_accessor :raw_info

			# def self.humanid_button
			# 	#see https://docs.human-id.org/web-sdk-integration-guide
			# 	%Q{<a href="#{options.local_sign_up_url}">
    		# 		<img src="#{options.login_button_path}" alt="Anonymous Login with humanID" height="27"/>
			# 	<a>}
			# end

			def get_client_id
			#basic check for client_id
				return options.client_id unless options.client_id.nil?
				raise StandardError.new("Please set omniauth-humanid client id")
			end
			def get_client_secret
			#basic check for client_secret
				return options.client_secret unless options.client_secret.nil?
				raise StandardError.new("Set omniauth-humanid client secret")
			end
			def get_external_signup_uri
				uri = URI(options.external_signup_url.gsub('[HUMANID_VERSION]', options.humanid_version))
				query = [
					['lang', options.lang],
					['priority_country', options.priority_country]
				].select{|a| not a[1].nil?}
				uri.query = URI.encode_www_form(query)
				return uri
			end
			def request_phase_err(res)
				raise StandardError.new("Issue with the request phase of humanid omniauth, response from human id has code: #{res.code}, and body: #{res.body}")
			end
			#request phase
			def request_phase
				# In the humanid web-sdk-integration-guide, this would be the "[1] login" step. We need to get the redirect url
				# through a post request, and then send that to the user.

				#see more on Net::HTTP here:
				#	https://ruby-doc.org/stdlib-2.4.1/libdoc/net/http/rdoc/Net/HTTP.html

				#get uri
				uri = get_external_signup_uri
				Rails.logger.debug "HUMANID_OMNIAUTH URI: #{uri.to_s}"
				#make a post request (but dont send it yet)
				post_request = Net::HTTP::Post.new(uri)
				#set the headers as per docs.
				post_request['client-id'] = get_client_id
				post_request['client-secret'] = get_client_secret
				post_request['Content-Type'] = 'application/json'
				#send the request using a weirdly 🤷‍♂️ overcomplicated 🤷‍♂️ method 🤷‍♂️ and 🤷‍♂️ block 🤷‍♂️ blame Net::HTTP
				res = Net::HTTP.start(uri.hostname, uri.port, use_ssl: true){|http| http.request(post_request)}
				Rails.logger.info(res)
				if res.code == "200"
					body = JSON.parse(res.body)
					# they have five diffrent metrics for success:
					#	1. the response code 200
					#	2. body["success"] == true
					#	3. body["code"] == "OK"
					#	4. body["message"] == "success"
					#	5. body["data"]["webLoginUrl"] is actually there
					# I check 1 and 5 since the others seem supplimentary
					
					#get the redirect url
					Rails.logger.info(body)
					redirect_url = body.dig("data", "webLoginUrl")
					#check it, throw an error if nil or an int or something random.
					request_phase_err(res) unless redirect_url.kind_of? String
					#redirect (everything is working!)
					Rails.logger.info(redirect_url)
					redirect redirect_url
				else
					request_phase_err(res)
				end
			end
			#callback phase area

			def get_exchange_uri
				uri = URI(options.exchange_url.gsub('[HUMANID_VERSION]', options.humanid_version))
				return uri
			end

			def callback_phase
				#when the callback returns from humanID we still need to:
				#	1. verify it is humanID who sent this request
				#	2. get the uid and country code
				#this is done in the verify exchange token step in the humanID docs.

				#get the exchange_token from the humanID callback
				Rails.logger.info("CALLBACK PHASE")
				exchange_token = request.params['et']
				Rails.logger.info("EXCHANGE TOKEN: #{exchange_token}")
				
				#create the request (as per the humanID docs)
				uri = get_exchange_uri
				post_request = Net::HTTP::Post.new(uri)
				post_request['client-id'] = get_client_id
				post_request['client-secret'] = get_client_secret
				post_request['Content-Type'] = 'application/json'
				post_request.body = {"exchangeToken" => exchange_token}.to_json
				#send the request, get the response.
				res = Net::HTTP.start(uri.hostname, uri.port, use_ssl: true){|http| http.request(post_request)}
				Rails.logger.info("RESPONSE: #{res}")
				if res.code == "200"
					self.raw_info = JSON.parse(res.body)
					Rails.logger.info("raw: #{raw_info}")
					super
				else
					str = "Issue with the callback_phase of humanid omniauth, response from human id has code: #{res.code}, and body: #{res.body}"
					Rails.logger.error str
					raise StandardError.new(str)
				end
			end
			
			#not a method? Some DSL magic that is not explained in docs. Just looked at other projects and they did something like this so
			info do
				raw_info['data']
			end
			uid do
				raw_info['data']['userAppId']
			end
		end
	end
end
